tests/common/plugins/conditional_mark/__init__.py

```diff
@@ -402,31 +402,53 @@ def load_basic_facts(session):
     return results
 
 
-def find_longest_matches(nodeid, conditions):
-    """Find the longest matches of the given test case name in the conditions list.
-
-    This is similar to longest prefix match in routing table. The longest match takes precedence.
+def find_all_matches(nodeid, conditions):
+    """Find all matches of the given test case name in the conditions list.
 
     Args:
         nodeid (str): Full test case name
         conditions (list): List of conditions
 
     Returns:
-        str: Longest match test case name or None if not found
+        list: All match test case name or None if not found
     """
-    longest_matches = []
+    all_matches = []
     max_length = -1
+    conditional_marks = {}
+    matches = []
+
     for condition in conditions:
         # condition is a dict which has only one item, so we use condition.keys()[0] to get its key.
         if nodeid.startswith(list(condition.keys())[0]):
-            length = len(list(condition.keys())[0])
-            if length > max_length:
-                max_length = length
-                longest_matches = []
-                longest_matches.append(condition)
-            elif length == max_length:
-                longest_matches.append(condition)
-    return longest_matches
+            all_matches.append(condition)
+
+    for match in all_matches:
+        case_starting_substring = list(match.keys())[0]
+        length = len(case_starting_substring)
+        marks = match[case_starting_substring].keys()
+        for mark in marks:
+            if mark in conditional_marks:
+                if length > max_length:
```

Comment:
Shouldn't the `max_length` be a per-mark attribute?
Otherwise if you had the following:
```
package:
   skip:
      condition:
         -<conditionA>
   xfail:
      condition:
         -<conditionB>

package/test.py::testName:
   skip:
      condition:
         -<conditionC>

package/test.py:
   xfail:
      condition:
         -<conditionD>
```

When the match `package/test.py::testName` is processed against the `skip` mark it will cause the `max_length` to be set to `len('package/test.py::testName')` but then later when `package/test.py` is processed for `xfail` it won't be added because it's incorrectly comparing against the `max_length` of the `skip` mark.

Reply:
Hi, @arista-nwolfe, in our yaml file of conditional mark, we set all conditions in alphabet order. So we will check the `case_starting_substring` from short to long.

Reply:
> Hi, @arista-nwolfe, in our yaml file of conditional mark, we set all conditions in alphabet order. So we will check the `case_starting_substring` from short to long.

Can we rely on the yaml file being populated that way? If we are then you technically don't even need a `if length > max_length:` check since the latter processed match will always be the same length or longer than the previous matches.

Reply:
> > Hi, @arista-nwolfe, in our yaml file of conditional mark, we set all conditions in alphabet order. So we will check the `case_starting_substring` from short to long.
> 
> Can we rely on the yaml file being populated that way? If we are then you technically don't even need a `if length >= max_length:` check since the latter processed match will always be the same length or longer than the previous matches.

Yes, we have a pre-commit checker to check.